src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/TaintedDataAnalysis/WebInputSources.cs

```diff
@@ -12,13 +16,57 @@ internal static class WebInputSources
         /// </summary>
         public static ImmutableHashSet<SourceInfo> SourceInfos { get; }
 
+        private static readonly BoundedCacheWithFactory<Compilation, ConcurrentDictionary<ISymbol, bool>> s_sourceArgumentCacheByCompilation = new BoundedCacheWithFactory<Compilation, ConcurrentDictionary<ISymbol, bool>>();
```

Comment:
Instead of caching by the `IMethodSymbol` of the potential action method, I think it'd be better to cache the `INamedTypeSymbol` of the potential controller. That'd be a smaller cache and get more cache hits.

Then it's probably okay to check if the `IMethodSymbol` is an action method multiple times without caching, because at that point at least we'd know we're inside a controller.

Also add doc comments for what all the keys and values are.

Reply:
I have added a second cache to address full recompilation too. Let me know if you think it is overkill.